mewarpx/mewarpx/mwxrun.py

```diff
@@ -174,5 +181,53 @@ def get_npart_species_dict(self):
 
         return npart_dict
 
+    def get_rho_grid(self):
+        """Get rho segments on the grid for each tile of each processor.
+
+        Returns:
+            A list of numpy arrays, the list has an array for every tile and
+            each array has dimensions given by the number of cells in that tile.
+        """
+        return _libwarpx.get_mesh_charge_density_fp(self.lev)
+
+    def get_gathered_rho_grid(self):
+        """Get the full rho on the grid on the root processor.
+
+        Returns:
+            A list with only 1 element - a numpy array with rho on the full
+            domain.
+        """
+        return _libwarpx.get_gathered_charge_density_fp(self.lev)
+
+    def get_phi_grid(self):
+        """Get phi segments on the grid for each tile of each processor.
+
+        Returns:
+            A list of numpy arrays, the list has an array for every tile and
+            each array has dimensions given by the number of cells in that tile.
+        """
+        return _libwarpx.get_mesh_phi_fp(self.lev)
+
+    def get_gathered_phi_grid(self):
+        """Get the full phi on the grid on the root processor.
+
+        Returns:
+            A list with only 1 element - a numpy array with phi on the full
+            domain.
+        """
+        return _libwarpx.get_gathered_phi_fp(self.lev)
+
+    def set_phi_grid(self, phi_data):
+        """Sets phi segments on the grid to input phi data"""
+        # only proc 0 has the gathered phi grid so only it should set
+        # the phi grid
+        if self.me == 0:
+            # get phi multifab from warpx
+            phi_ptr = _libwarpx.get_pointer_full_phi_fp(self.lev)
+            try:
+                phi_ptr[0][:] = phi_data
+            except ValueError:
```

Comment:
You should check text of the ValueError to make sure this is indeed the error. Then, you should probably print what you believe is the problem and re-raise the error (just write `raise` on its own line), rather than SystemExit.

(Reasoning: In principle, you might have a python script that runs a bunch of WarpX simulations. It would have its own error handler that would mark a run as failed if it raises an error, but we wouldn't want this error to quit the script running all the simulations. So raising is almost always the better option - if nothing catches the error it'll cause program exit anyway.)

Reply:
I'll change this now, but for learning sake @KZhu-ME see the above.

Reply:
I'll keep that in mind for the future. Thanks for the heads up!